src/adc.rs

```diff
@@ -0,0 +1,562 @@
+    gpiod::PD14<Analog> => 11_u8,
+);
+
+// # ADC4 Pin/Channel mapping
+// ## f303
+
+#[cfg(any(
+    feature = "stm32f303xb",
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+))]
+adc_pins!(ADC4,
+    gpioe::PE14<Analog> => 1_u8,
+    gpioe::PE15<Analog> => 2_u8,
+    gpiob::PB12<Analog> => 3_u8,
+    gpiob::PB14<Analog> => 4_u8,
+    gpiob::PB15<Analog> => 5_u8,
+    gpiob::PB8<Analog> => 12_u8,
+    gpiob::PB9<Analog> => 13_u8,
+    // Shared channels (i.e. ADC34_INx)
+    gpioe::PE8<Analog> => 6_u8,
+    gpiod::PD10<Analog> => 7_u8,
+    gpiod::PD11<Analog> => 8_u8,
+    gpiod::PD12<Analog> => 9_u8,
+    gpiod::PD13<Analog> => 10_u8,
+    gpiod::PD14<Analog> => 11_u8,
+);
+
+/// Abstract implementation of ADC functionality
+///
+/// Do not use directly. See adc12_hal for a applicable Macro.
+/// TODO: Extend/generalize beyond f303
+macro_rules! adc_hal {
+    ($(
+            $ADC:ident: ($adcx:ident, $ADC_COMMON:ident),
+    )+) => {
+        $(
+            impl Adc<$ADC> {
+
+                /// Init a new ADC
+                ///
+                /// Enables the clock, performs a calibration and enables the ADC
+                pub fn $adcx(
+                    rb: $ADC,
+                    adc_common : &mut $ADC_COMMON,
+                    ahb: &mut AHB,
+                    ckmode: CKMODE,
+                    clocks: Clocks,
+                ) -> Option<Self> {
+                    let mut this_adc = Self {
+                        rb,
+                        clocks,
+                        ckmode,
+                        operation_mode: None,
+                    };
+                    if !(this_adc.clocks_welldefined(clocks)) {
+                        return None;
+                    }
+                    if !(this_adc.enable_clock(ahb, adc_common)){
+                        return None;
+                    }
+                    this_adc.set_align(Align::default());
+                    this_adc.calibrate();
+                    // ADEN bit cannot be set during ADCAL=1
+                    // and 4 ADC clock cycle after the ADCAL
+                    // bit is cleared by hardware
+                    this_adc.wait_adc_clk_cycles(4);
+                    this_adc.enable();
+                    Some(this_adc)
+                }
+
+                /// Software can use CKMODE::SYNCDIV1 only if
+                /// hclk and sysclk are the same. (see reference manual 15.3.3)
+                fn clocks_welldefined(&self, clocks: Clocks) -> bool {
+                    !(self.ckmode == CKMODE::SYNCDIV1 && !(clocks.hclk().0 == clocks.sysclk().0))
+                }
+
+                /// sets up adc in one shot mode for a single channel
+                pub fn setup_oneshot(&mut self) {
+                    // stop and clear overrun events
+                    self.rb.cr.modify(|_, w| w.adstp().set_bit());
+                    self.rb.isr.modify(|_, w| w.ovr().clear_bit());
```

Comment:
I think clearing the bit is wrong here. According to the RM: "OVR: ADC overrun. This bit is set by hardware when an overrun occurs on a regular channel, meaning that a new conversion has completed while the EOC flag was already set. **It is cleared by software writing 1 to it.**" So, to clear the overrun flag, you have to actually set this bit.

If registers have more descriptive methods than the generic `set_bit` and `clear_bit`, it is usually a good idea to use these. It prevents bugs like this one and also improve readability. You use `set_bit` and `clear_bit` pretty consistently here. I'd suggest converting them all to the descriptive methods, if possible.

Reply:
Thanks for spotting this. When I first wrote this module, the old PAC didn't provide the more descriptive methods. 
I'm all in with using them.